File: app/src/main/java/org/jellyfin/androidtv/ui/FloatSpinner.java
package org.jellyfin.androidtv.ui;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import org.jellyfin.androidtv.R;

public class FloatSpinner extends FrameLayout {
    float mValue = 1;
    float mIncrement = .1f;
    TextView mTextValue;
    ValueChangedListener<Float> mValueChangedListener;


    public FloatSpinner(Context context, ValueChangedListener<Float> listener) {
        super(context);
        mValueChangedListener = listener;
        init(context);
    }

    public FloatSpinner(Context context) {
        super(context);
        init(context);
    }

    public FloatSpinner(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        LayoutInflater inflater = LayoutInflater.from(context);
        View v = inflater.inflate(R.layout.number_spinner, this, true);
        if (!isInEditMode()) {
            mTextValue = (TextView) v.findViewById(R.id.txtValue);
            (v.findViewById(R.id.btnIncrease)).setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    setValue(mValue + mIncrement);
                }
            });
            (v.findViewById(R.id.btnDecrease)).setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    setValue(mValue-mIncrement);
                }
            });
        }
    }

    public void setOnChangeListener(ValueChangedListener<Float> listener) {
        mValueChangedListener = listener;
    }

    public void setValue(float value) {
        mValue = value;
        mTextValue.setText(String.format(getResources().getConfiguration().locale, "%.1f", mValue));
        if (mValueChangedListener != null) {
            mValueChangedListener.onValueChanged(value);
        }
    }

    public float getValue() {
        return mValue;
    }

    public void setIncrement(float value) {
        mIncrement = value;
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/FloatSpinner.kt
package org.jellyfin.androidtv.ui

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import org.jellyfin.androidtv.ui.ValueChangedListener
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import org.jellyfin.androidtv.R
import org.jellyfin.androidtv.databinding.NumberSpinnerBinding

class FloatSpinner @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
    var minValue: Float? = null,
    var maxValue: Float? = null,
	private var changeListener: ((value: Float) -> Unit)? = null,
) : FrameLayout(context, attrs) {
    private var currentValue = 1f
    private var increment = .1f
	private val binding: NumberSpinnerBinding = NumberSpinnerBinding.inflate(LayoutInflater.from(context), this, true)

    init {
        if (!isInEditMode) {
        	binding.btnIncrease.setOnClickListener {
                value = if (maxValue == null || currentValue + increment < maxValue!!) {
                    currentValue + increment
                } else {
                    maxValue
                }!!
        	}
        	binding.btnDecrease.setOnClickListener {
                value = if (minValue == null || currentValue - increment > minValue!!) {
                    currentValue - increment
                } else {
                    minValue
                }!!
        	}
        }
    }

    fun setOnChangeListener(listener: ((value: Float) -> Unit)?) {
		changeListener = listener
    }

    var value: Float
        get() = currentValue
        set(value) {
			currentValue = value
			binding.txtValue.text = String.format(resources.configuration.locale, "%.1f", currentValue)
			changeListener?.invoke(value)
        }

    fun setIncrement(value: Float) {
        increment = value
    }
}
